features/environment.py

```diff
@@ -13,16 +14,13 @@
 from validation_results import ValidationOutcome, ValidationOutcomeCode, OutcomeSeverity
 from main import ExecutionMode
 
-model_cache = {}
-def read_model(fn, pure):
-    if cached := model_cache.get(fn):
-        return cached
-    if pure:
-        model_cache[(fn, pure)] = ifcopenshell.simple_spf.open(fn)
-    else:
-        model_cache[(fn, pure)] = ifcopenshell.open(fn)
-    return model_cache[(fn, pure)]
+@functools.cache
+def _open_model(fn, pure):
+    return (ifcopenshell.simple_spf.open(fn)
+            if pure else ifcopenshell.open(fn))
 
+def read_model(fn, pure):
+    return _open_model(fn, pure)
```

Comment:
```suggestion
@functools.cache
def read_model(fn, pure):
    return (ifcopenshell.simple_spf.open(fn)
            if pure else ifcopenshell.open(fn))
```

I don't think any extra indirection is needed. As long as the functools.cache is a top-level item in the namespace so that the decorator is applied only once.